### ISSUE TITLE
Instructions for cluster mode?

### ISSUE THREAD

**Issue by kcalmond** [pasky/pachi#140]
After reviewing the repo an issues I don't find anything that describes how to deploy Pachi into a cluster of nodes. 

Specifically, as a hobby project I have created a Raspberry Pi4B cluster running Ubuntu 18.04 on each node, as a k3S cluster (Kubernetes). I'm looking for interesting apps/services to run on it. Since I'm a go enthusiast I thought running a gobot server on the cluster would be interesting project. So deployment mode here would be clustered Pachi, in containers.  Any ideas?

**Comment by lemonsqueeze** [comment]
Hi,
Sure, depending on what you have in mind there are lots of things you could do if you put these Raspberry Pis to work. Distributed Pachi will allow you to make a stronger (or faster) engine. Or if you're interested in making it play online you could power many instances as opposed to just one or two.

Distributed mode needs some work at the moment. Looks like I broke it with last gtp layer upgrade, but should be easy to fix. As for strength, expect to spend huge amounts of cpu to gain just one stone. For example at one point I was running Pachi on kgs with 180k playouts, it was just one stone stronger than now with 5-15k playouts. Patterns have improved since so it might be stronger now though. [distributed/distributed.c](https://github.com/pasky/pachi/blob/master/distributed/distributed.c) has instructions on how to set it up. Basically you start it with:
```
pachi -e distributed slave_port=1234          # master node
pachi -e uct -g masterhost:1234 slave         # slaves
```

For online play with multiple instances (several bots each playing their own game) there is infrastructure you can use. kgs is easy but not very flexible: you setup one instance/account on each raspberry pi (maybe even two with FIFO lock enabled so they don't take the cpu at the same time). ogs lets you load-balance so you could build something where you accept as many games as you have cpu power for and dispatch game commands to each node. Fox also has a gtp client and I don't think Pachi is playing there, so could be an idea also, this server is huge !

**Comment by lemonsqueeze** [comment]
Here's the fix for the distributed engine: PR #141
Should work fine with it now.

**Comment by kcalmond** [comment]
@lemonsqueeze My process for trying this out is:
dow

1. nload/run single node pachi on an Ubuntu host (to make sure I know how to run it and can connected a GTP client to it)
  Q: which GTP client should I use to smoke test?
3. Repeat #1 in docker - create dockerfile, image etc
4. Test distributed mode in docker (run multiple containers on same host)
5. Test distributed mode in k8s (deployment, helm chart etc.)
6. ? k8s horizo

ntal pod autoscaler for new game/session requests?

**Comment by lemonsqueeze** [comment]
Yes, you will need to build Pachi yourself for this, binary releases don't have the distributed engine built in.
* First see if you can get vanilla Pachi to build (following instructions in [INSTALL.md](https://github.com/pasky/pachi/blob/master/INSTALL.md))
by default it won't have the distributed engine (or the fix) either but it's a good start.
* To get the fix, remove pachi directory and start again, but this time do:
    ```
    git clone 'https://github.com/lemonsqueeze/pachi'
    cd pachi
    git checkout distributed_fix
    ```
* To build the distributed engine, edit the Makefile and uncomment:
    ```
    DISTRIBUTED=1
    NETWORK=1
    ```
    then
    ```
    make clean
    make -j4
    ```
    Now you should have a working distributed engine.

**Comment by lemonsqueeze** [comment]
If you're new to Pachi i'd suggest playing with it in normal mode first to get familiar with it. You can bring it up in a gtp client like gogui or just type the commands directly:
```
$ ./pachi
boardsize 19
clear_board
genmove b
```
should make it generate a move for example.

Then you can try the distributed engine on the same machine:
```
./pachi -t =4000  -e distributed slave_port=1234
```
And in another terminal:
```
./pachi  -g localhost:1234 slave
```
Now if you type commands in the first window it should get the slave to generate a move.

**Comment by kcalmond** [comment]
Thx again @lemonsqueeze 👍 .  The `make` ran fine on my ubuntu 18.04 system. `make install` too, however I had to sudo that command.  When running `make install-data` I got the warnings below on stdout. As a guess I tried `make install-data` again after unzipping the detlef54 data files into /usr/local/share/pachi. Output from `make install-data` produced same output:

 ```
~/git/pachi  master ?1  sudo make install-data
/usr/bin/install -d /usr/share/pachi-go
/usr/bin/install patterns_mm.gamma /usr/share/pachi-go/
/usr/bin/install patterns_mm.spat /usr/share/pachi-go/
WARNING: book.dat datafile is missing
WARNING: golast19.prototxt datafile is missing
WARNING: golast.trained datafile is missing
/usr/bin/install joseki19.gtp /usr/share/pachi-go/
```

**Comment by lemonsqueeze** [comment]
Yes that looks good, well done !
By the way you don't really need to install anything (the sudo part) at this point. If you copy the data files in the directory where you built pachi you can just run everything from there:
 `./pachi` or the full path, say `~/src/pachi/pachi` if the build is in ~/src/pachi.
But having them in `/usr/share/pachi-go` won't hurt either.

**Comment by lemonsqueeze** [comment]
Distributed Engine Fix PR #141 has been merged, so now you can clone main repository to play with distributed engine :

 `git clone https://github.com/pasky/pachi`
